kong/db/strategies/off/init.lua

```diff
@@ -0,0 +1,150 @@
+local declarative_config = require("kong.db.schema.others.declarative_config")
+
+
+local kong = kong
+local fmt = string.format
+local tostring = tostring
+local tonumber = tonumber
+local encode_base64 = ngx.encode_base64
+local decode_base64 = ngx.decode_base64
+
+
+local off = {}
+
+
+local _mt = {}
+_mt.__index = _mt
+
+
+local function page_for_key(self, key, size, offset)
+  if offset then
+    local token = decode_base64(offset)
+    if not token then
+      return nil, self.errors:invalid_offset(offset, "bad base64 encoding")
+    end
+
+    token = tonumber(token)
+    if not token then
+      return nil, self.errors:invalid_offset(offset, "invalid offset")
+    end
+
+    offset = token
+
+  else
+    offset = 1
+  end
+
+  if not kong.cache then
+    return {}
+  end
+
+  local list = kong.cache:get(key, nil, function()
+    return {}
```

Comment:
Should we return `nil, err` here? So that we do not cache the return value, and we actually enter the below `if not list` branch?

Reply:
no, this is meant to return `{}`. If you are trying to paginate on an entity that has no elements (e.g. you have no consumers and you hit `/services/mysrv/routes` and `mysrv` has no routes), the cache will be empty with no `list` entry for it, and the correct answer is an empty list.